base/twiceprecision.jl

```diff
@@ -254,7 +254,10 @@ nbitslen(::Type{T}, len, offset) where {T<:IEEEFloat} =
     min(cld(precision(T), 2), nbitslen(len, offset))
 # The +1 here is for safety, because the precision of the significand
 # is 1 bit higher than the number that are explicitly stored.
-nbitslen(len, offset) = len < 2 ? 0 : ceil(Int, log2(max(offset-1, len-offset))) + 1
+nbitslen(len, offset) = len < 2 ? 0 : ceil_log2(max(offset-1, len-offset)) + 1
+
+ceil_log2(n::Integer) = ceil(Int, log2(n))
+ceil_log2(n::T) where {T<:BitSigned} = 8 * sizeof(T) - leading_zeros(n - T(1))
```

Comment:
This is a common pattern (e.g. [used in sorting](https://github.com/JuliaLang/julia/blob/1b2d37b7f77c17fab0d781def342461fe5be35da/base/sort.jl#L830))

To avoid duplication throughout base, I think it makes sense to define in base/int.jl and document it (#47523)

Reply:
The same pattern is used several times also in https://github.com/JuliaLang/julia/blob/0a65af86ad2b14d28c85d3afce161cdf24a93f66/base/intfuncs.jl. It'd be good to consolidate it, instead of having it implemented several times inconsistently

Reply:
Fine. That PR should also just replace the uses (and documentation) in base/twiceprecision.jl with the new function.

Reply:
That should be a pretty simple diff
```suggestion
ceil_log2(n::Integer) = ceil(Int, log2(n))
ceil_log2(n::BitSigned) = ndigits0z2(n - oneunit(n))
```

Reply:
Wait you propose that we make two new functions, `ceil_log2` *and* `ndigits0z2`?

Reply:
Both PRs stand on their own, but merging them seems reasonable.

Reply:
Hmm. It still seems very surprising to me to introduce two brand new functions to do that same thing.

You can of course rebase https://github.com/JuliaLang/julia/pull/47523 on top of this & delete the name. One of them, either of them, it's internal and deserves an ugly name, but just one name.